data-analytics/cloud-composer-examples/composer_dataflow_examples/simple_load_dag.py

```diff
@@ -0,0 +1,147 @@
+DS_TAG = '{{ ds }}'
+DATAFLOW_FILE = os.path.join(
+    configuration.get('core', 'dags_folder'), 'dataflow', 'process_delimited.py')
+
+# The following additional Airflow variables should be set:
+# gcp_project:         Google Cloud Platform project id.
+# gcp_temp_location:   Google Cloud Storage location to use for Dataflow temp location.
+# email:               Email address to send failure notifications.
+DEFAULT_DAG_ARGS = {
+    'start_date': YESTERDAY,
+    'email': models.Variable.get('email'),
+    'email_on_failure': True,
+    'email_on_retry': False,
+    'retries': 0,
+    'project_id': models.Variable.get('gcp_project'),
+    'dataflow_default_options': {
+        'project': models.Variable.get('gcp_project'),
+        'temp_location': models.Variable.get('gcp_temp_location'),
+        'runner': 'DataflowRunner'
+    }
+}
+
+
+def move_to_completion_bucket(target_bucket, target_infix, **kwargs):
+    """A utility method to move an object to a target location in GCS."""
+    # Here we establish a connection hook to GoogleCloudStorage.
+    # Google Cloud Composer automatically provides a google_cloud_storage_default
+    # connection id that is used by this hook.
+    conn = gcs_hook.GoogleCloudStorageHook()
+
+    # The external trigger (Google Cloud Function) that initiates this DAG
+    # provides a dag_run.conf dictionary with event attributes that specify
+    # the information about the GCS object that triggered this DAG.
+    # We extract the bucket and object name from this dictionary.
+    source_bucket = kwargs['dag_run'].conf['bucket']
+    source_object = kwargs['dag_run'].conf['name']
+    completion_ds = kwargs['ds']
+
+    target_object = os.path.join(target_infix, completion_ds, source_object)
+
+    logging.info('Coping %s to %s',
+                 os.path.join(source_bucket, source_object),
+                 os.path.join(target_bucket, target_object))
+    conn.copy(source_bucket, source_object, target_bucket, target_object)
+
+    logging.info('Deleting %s',
+                 os.path.join(source_bucket, source_object))
+    conn.delete(source_bucket, source_object)
+
+
+# Setting schedule_interval to None as this DAG is externally trigger by a Cloud Function.
+# The following Airflow variables should be set for this DAG to function:
+# bq_output_table: BigQuery table that should be used as the target for
+#                  Dataflow in <dataset>.<tablename> format.
+#                  e.g. lake.usa_names
+# input_field_names: Comma separated field names for the delimited input file.
+#                  e.g. state,gender,year,name,number,created_date
+with models.DAG(dag_id='GcsToBigQueryTriggered',
+                description='A DAG triggered by an external Cloud Function',
+                schedule_interval=None, default_args=DEFAULT_DAG_ARGS) as dag:
+    # Args required for the Dataflow job.
+    job_args = {
+        'input': 'gs://{{ dag_run.conf.get("bucket") }}/{{ dag_run.conf.get("name") }}',
```

Comment:
Would it be better here to use `'input': 'gs://{{ dag_run.conf["bucket"] }}/{{ dag_run.conf["name"] }}',` to give the user a KeyError  if the `dag_run.conf` doesn't contain these keys? For example, if the user manually triggers this DAG from the UI (therefore the `dag_run.conf` would not have the `"bucket"` or `"name"` keys) then they would kick off a DataFlow job that is DOA. I could see using the `.get` method if your wanted to add a default value that might lead to a successful DataFlow job but it seems we should error out before spinning up DataFlow workers and waiting for DataFlow to handle the error.